deepset_cloud_sdk/workflows/pipeline_client/pipeline_service.py

```diff
@@ -0,0 +1,257 @@
+"""Pipeline publishing service for deepset Cloud SDK."""
+from contextlib import asynccontextmanager
+from typing import Any, AsyncGenerator, Dict, Protocol, runtime_checkable
+
+import structlog
+
+from deepset_cloud_sdk._api.config import DEFAULT_WORKSPACE_NAME, CommonConfig
+from deepset_cloud_sdk._api.deepset_cloud_api import DeepsetCloudAPI
+from deepset_cloud_sdk.workflows.pipeline_client.models import (
+    PipelineInputs,
+    PipelineOutputs,
+    PipelineType,
+    PublishConfig,
+)
+from deepset_cloud_sdk.workflows.user_facing_docs.pipeline_service_docs import (
+    PipelineServiceDocs,
+)
+
+logger = structlog.get_logger(__name__)
+
+
+@runtime_checkable
+class PipelineProtocol(Protocol):
+    """Protocol defining the required methods for a pipeline object."""
+
+    def dumps(self) -> str:
+        """Convert the pipeline to a YAML string.
+
+        :return: YAML string representation of the pipeline
+        """
+        ...
+
+    def to_dict(self) -> Dict[str, Any]:
+        """Convert the pipeline to a dictionary representation.
+
+        :return: Dictionary representation of the pipeline
+        """
+        ...
+
+    def add_component(self, name: str, instance: Any) -> None:
+        """Add a component to the pipeline.
+
+        :param name: Name of the component
+        :param instance: Component instance to add
+        """
+        ...
+
+    def connect(self, sender: str, receiver: str) -> "PipelineProtocol":
+        """Connect two components in the pipeline.
+
+        :param sender: Component name or component.connection_name that sends data
+        :param receiver: Component name or component.connection_name that receives data
+        :return: The pipeline instance for method chaining
+        """
+        ...
```

Comment:
do we need these? `dumps` is the the only one we care about, right? we cloud also check the parent class which would require us knowing less about the implementation details of haystack and be more robust

Reply:
yeah. When it's only one method your checking against, a simple `gettattr` is probably the simpler option 👍🏻